joeynmt/model.py

```diff
@@ -159,21 +164,33 @@ def run_batch(self, batch: Batch, max_output_length: int, beam_size: int,
 
         # greedy decoding
         if beam_size == 0:
-            stacked_output, stacked_attention_scores = greedy(
-                encoder_hidden=encoder_hidden, encoder_output=encoder_output,
-                src_mask=batch.src_mask, embed=self.trg_embed,
-                bos_index=self.bos_index, decoder=self.decoder,
-                max_output_length=max_output_length)
+
+            # greedy decoding for Transformer models
+            if isinstance(self.decoder, TransformerDecoder):
+                stacked_output, stacked_attention_scores = transformer_greedy(
+                    encoder_output=encoder_output,
+                    src_mask=batch.src_mask, embed=self.trg_embed,
+                    bos_index=self.bos_index, decoder=self.decoder,
```

Comment:
Idea: let's keep one global `greedy` function and make a `recurrent_greedy` vs. `transformer_greedy` distinction within that global function, based on the `decoder` object that is passed to it. That way it's wrapped a bit more nicely.

`stacked_output, stacked_attention_scores = greedy(
                encoder_hidden=encoder_hidden, encoder_output=encoder_output,
                src_mask=batch.src_mask, embed=self.trg_embed,
                bos_index=self.bos_index, decoder=self.decoder,
                max_output_length=max_output_length)`

and then in `greedy`:
`if isinstance(decoder, TransformerDecoder):` ...